File: hacker_rank/statistics/interquartile_range.rb
num_els = gets.chomp
els = gets.chomp.split(' ')
els_count = gets.chomp.split(' ')
nums = []
els.each_with_index do |el, idx|
    els_count[idx].to_i.times { nums << el.to_i }
end
nums.sort!
first_half = nums.take(nums.size / 2)
second_half = nums.drop(nums.size.even? ? nums.size / 2 : nums.size / 2 + 1)
q1 = [first_half[first_half.size / 2]]
q3 = [second_half[second_half.size / 2]]
q1 << first_half[first_half.size / 2 - 1] if first_half.size.even?
q3 << second_half[second_half.size / 2 - 1] if second_half.size.even?
one = q1.reduce(:+).to_f / q1.size
three = q3.reduce(:+).to_f / q3.size
puts (three - one).round(1)

File: hacker_rank/statistics/InterquartileRange.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	numEls, number, q1, q2, q3 := 0, 0, 0, 0, 0
	count := make(map[int]int)
	var nums []int
	var numbers []int
	fmt.Scanf("%d", &numEls)
	for i := 0; i < numEls; i++ {
		fmt.Scanf("%d", &number)
		nums = append(nums, number)
	}
	for i := 0; i < numEls; i++ {
		fmt.Scanf("%d", &number)
		for j := 0; j < number; j++ {
			numbers = append(numbers, nums[i])
		}
	}
	sort.Ints(numbers)
	if len(numbers)%2 == 0 {
		q2 = (numbers[len(numbers)/2] + numbers[len(numbers)/2-1]) / 2
		if (len(numbers)/2)%2 == 0 {
			q1 = (numbers[(len(numbers)/2)/2] + numbers[(len(numbers)/2)/2-1]) / 2
			q3 = (numbers[(len(numbers)/2)/2*3] + numbers[(len(numbers)/2)/2*3-1]) / 2
		} else {
			q1 = numbers[(len(numbers)/2)/2]
			q3 = numbers[((len(numbers)/2)/2)*3]
		}
	} else {
		q2 = numbers[len(numbers)/2]
		if (len(numbers)/2)%2 == 0 {
			q1 = (numbers[(len(numbers)/2)/2] + numbers[(len(numbers)/2)/2-1]) / 2
			q3 = (numbers[(len(numbers)/2)/2*3] + numbers[(len(numbers)/2)/2*3-1]) / 2
		} else {
			q1 = numbers[(len(numbers)/2)/2]
			q3 = numbers[((len(numbers)/2)/2)*3]
		}
	}
	fmt.Println(q3 - q1)

}
